File: src/java/magma/api/option/Option.java
package magma.api.option;

import magma.api.Tuple;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public interface Option<T> {
    <R> Option<R> map(Function<T, R> mapper);

    T orElseGet(Supplier<T> other);

    T orElse(T other);

    <R> Option<R> flatMap(Function<T, Option<R>> mapper);

    void ifPresent(Consumer<T> consumer);

    Option<T> or(Supplier<Option<T>> other);

    boolean isPresent();

    Option<T> filter(Predicate<T> predicate);

    boolean isEmpty();

    Tuple<Boolean, T> toTuple(T other);
}

File: src/c/magma/api/option/Option.c
struct Option<T>{
	<R>Option<R> map(Tuple<any*, R (*)(T)> mapper);
	T orElseGet(Tuple<any*, T (*)()> other);
	struct Impl{
		struct Impl new(){
			struct Impl this;
			return this;
		}
	}
}